### ISSUE TITLE
Support baij and aij as separate matrix types

### ISSUE THREAD

**Issue by florianwechsung** [firedrakeproject/firedrake#914]
Hi, 

I am trying to solve the linear elasticity equations using the GAMG preconditioner. However, while my code works when I use hypre, is doesn't like gamg very much. The following simplified examples using the laplacian illustrate the problem:

```
from firedrake import *
parameters["pyop2_options"]["block_sparse"] = False
mesh = UnitSquareMesh(20, 20)
V = VectorFunctionSpace(mesh, "CG", 1)
u = TrialFunction(V)
v = TestFunction(V)

a = inner(grad(u), grad(v))*dx
L = (v[0]+v[1])*dx

u = Function(V)
solve(a == L, u, solver_parameters={"ksp_type": "cg", "mat_type" : "aij", "pc_type": "gamg"})
```

This fails with PETSc telling me `Require AIJ matrix type`. However, the following two examples work.

When I use hypre instead of gamg:

```
from firedrake import *
parameters["pyop2_options"]["block_sparse"] = False
mesh = UnitSquareMesh(20, 20)
V = VectorFunctionSpace(mesh, "CG", 1)
u = TrialFunction(V)
v = TestFunction(V)

a = inner(grad(u), grad(v))*dx
L = (v[0]+v[1])*dx

u = Function(V)
solve(a == L, u, solver_parameters={"ksp_type": "cg", "mat_type" : "aij", "pc_type": "hypre"})
```

Or when I solve in 1D instead of 2D:

```
from firedrake import *
parameters["pyop2_options"]["block_sparse"] = False
mesh = UnitSquareMesh(20, 20)
V = FunctionSpace(mesh, "CG", 1)
u = TrialFunction(V)
v = TestFunction(V)

a = inner(grad(u), grad(v))*dx
L = v*dx

u = Function(V)
solve(a == L, u, solver_parameters={"ksp_type": "cg", "mat_type" : "aij", "pc_type": "gamg"})
```

I was told on the slack channel that `parameters["pyop2_options"]["block_sparse"] = False` should result in a matrix of type aij for PETSc, but this doesn't seem to happen here.

**Comment by wence-** [comment]
Sorry, I lied to you about the name of the option, it should be `"block_sparsity"`.  Suggestions for how to make this more natural to control from Firedrake welcome.  We currently support:

```
mat_type: {aij, matfree, nest}
```

Where `aij` actually means `aij` for scalar spaces and `baij` for vector/tensor-valued spaces (made with `VectorFunctionSpace` or `TensorFunctionSpace`.  Similarly, in a nested matrix, there may be a block that is actually `baij`, not `aij`.

One thing we could do is let you say:

`block_mat_type: {aij, baij}` and interpret that appropriately (even inside nested matrices).  Thoughts?

**Comment by wence-** [comment]
I note in passing that for gamg to work well for your problem you'll need to provide the near null modes.  We currently don't have an interface to do this.  You can pass null modes in with `nullspace=...` to the solve, but can't do the same for PETSc's `nearnullspace` concept.  This is pretty easy to add (one can mimic the current code for null space attachment) if you'd like to have a go.  However, the keyword argument list will then be getting really long for the `solve` call.  Should we think about how to rework this?

**Comment by florianwechsung** [comment]
Thank you, that solved it! Running into convergence problems now but I assume that's related to not telling GAMG about the null-space. 

Your suggestion regarding an extra option ` block_mat_type` sounds good to me.

**Comment by florianwechsung** [comment]
Yeah I'm happy to give that a go, the whole thing is not time critical for me (hypre performs reasonably well right now) so we can decide on an interface first.

**Comment by wence-** [comment]
So the place to start would be to add a `nearnullspace` keyword argument to solve, that takes a `VectorSpaceBasis` or `MixedVectorSpaceBasis` object and propagates it through to the assembled operators (in the same way as is currently done for null spaces).  Start in `firedrake/solving.py` and work down to `firedrake/variational_solver.py` and thence to `firedrake/solving_utils.py`.  You'll also need to update the `_apply` method of the null space objects (in `firedrake/nullspace.py`) to handle the nearnullspace case.

**Comment by wence-** [comment]
This was fixed in #948, so you can now independently specify "aij" or "baij" as appropriate.  See that PR for more details.
